In [826]:
# učitavanje biblioteka 
import pandas as pd

# 1. Učitavanje baze i prikaz podataka

In [827]:
data = pd.read_csv('03_bone_marrow_dataset.csv')
print(data.shape)

(187, 37)


In [828]:
# brišemo razmak na kraju naziva atributa (lakše je za kasniju analizu)
for attribute in data.columns:
    if(attribute[-1] == ' '):
        data.rename(columns = {attribute:attribute[0:-1]}, inplace = True)
attributes = data.columns
data

Recipientgender  Stemcellsource  Donorage numeric  Donorage35  IIIV  \
0                  1               1         22.830137           0     1   
1                  1               0         23.342466           0     1   
2                  1               0         26.394521           0     1   
3                  0               0         39.684932           1     1   
4                  0               1         33.358904           0     0   
..               ...             ...               ...         ...   ...   
182                1               1         37.575342           1     1   
183                0               1         22.895890           0     0   
184                0               1         27.347945           0     1   
185                1               1         27.780822           0     1   
186                1               1         55.553425           1     1   

     Gendermatch  DonorABO RecipientABO RecipientRh ABOmatch  ... extcGvHD  \
0              0         1            1           1        0  ...        1   
1              0        -1           -1           1        0  ...        1   
2              0        -1           -1           1        0  ...        1   
3              0         1            2           1        1  ...        ?   
4              0         1            2           0        1  ...        1   
..           ...       ...          ...         ...      ...  ...      ...   
182            0         1            1           0        0  ...        1   
183            0         1            0           1        1  ...        1   
184            0         1           -1           1        1  ...        1   
185            0         1            0           1        1  ...        0   
186            0         1            2           1        1  ...        1   

    CD34kgx10d6 numeric CD3dCD34 numeric CD3dkgx10d8 numeric  \
0                  7.20          1.33876                5.38   
1                  4.50        11.078295                0.41   
2                  7.94         19.01323                0.42   
3                  4.25        29.481647                0.14   
4                 51.85         3.972255               13.05   
..                  ...              ...                 ...   
182               11.08          2.52275                4.39   
183                4.64         1.038858                4.47   
184                7.73         1.635559                4.73   
185               15.41          8.07777                1.91   
186                9.91         0.948135               10.45   

     Rbodymass numeric  ANCrecovery numeric  PLTrecovery numeric  \
0                   35                   19                   51   
1                 20.6                   16                   37   
2                 23.4                   23                   20   
3                   50                   23                   29   
4                    9                   14                   14   
..                 ...                  ...                  ...   
182                 44                   15                   22   
183               44.5                   12                   30   
184                 33                   16                   16   
185                 24                   13                   14   
186                 37                   18                   20   

     time_to_aGvHD_III_IV numeric  survival_time numeric  \
0                              32                    999   
1                         1000000                    163   
2                         1000000                    435   
3                              19                     53   
4                         1000000                   2043   
..                            ...                    ...   
182                            16                    385   
183                       1000000                    634   
184                       100

Podaci koji se nalaze u bazi:       

1. Recipientgender - pol pacijenta: 0 ili 1    
2. Stemcellsource - izvor ćelija (kvr ili koštana srž): 0 ili 1
3. Donorage numeric - broj godina donora: float num 18.646575 - 55.553425
4. Donorage35 - da li je starost donora manja od 35 godina
5. IIIV - razvoj bolesti transplantata 3. ili 4. stepena: 0 ili 1 
6. Gendermatch - kompatibilnost donora i promaoca po polu: 0 ili 1
7. DonorABO - krvna grupa donora: -1, 0, 1, 2
8. <font color=red>RecipientABO - krvna grupa pacijenta: -1, 0, 1, 2</font>
9. <font color=red>RecipientRh - da li pacijent ima Rh raktor: 0 ili 1</font>
10. <font color=red>ABOmatch - kompatibilnost pacijenta i donora po krvnoj grupi: 0 ili 1</font>
11. <font color=red>CMVstatus - kompatibilnost pacijenta i donora prema citomegalo virusu koji se može razviti (što je veća kompatibilnost to je lošije po pacijenta): num 0, 1, 2, 3 i ? </font>
12. <font color=red>DonorCMV - postojanje citomegalo virusne infekcije kod donora pre transplantacije: 0 ili 1 </font>
13. <font color=red>RecipientCMV - postojanje citomegalo virusne infekcije kod pacijenta pre transplantacije: 0 ili 1 </font>
14. Disease - vrsta bolesti: ALL, AML, chronic, nonmalignant, lymphoma
15. Riskgroup - rizična grupa: 0 ili 1
16. Txpostrelapse - da li je rađena druga transplantacija nakon što je prva prošla loše: 0 ili 1
17. Diseasegroup - da li je oboljenje maligno ili nije: 0 ili 1
18. HLAmatch - grupe antigena donora i primaoca: 0, 1, 2, 3
19. HLAmismatch - da li su se antigeni poklopili: 0 ili 1
20. <font color=red>Antigen - u koliko antigena postoji razlika između donora i pacijenta: -1, 0, 1, 2 i ?</font>
21. <font color=red>Alel - u koliko allela postoji razlika između donora i pacijenta: -1, 0, 1, 2, 3, ?</font>
22. HLAgrI - razlika između donora i primaoca u parametrima kao što su antigen, alel, HLA, itd: 0, 1, 2, 3, 4, 5, 6, 7
23. Recipientage numeric - starost pacijenta u trenutku transplantacije: 0.6 - 20.2
24. Recipientage10 - da li je starost pacijenta manja od 10 godina: 0 ili 1 
25. Recipientageint - godište pacijenta diskretizovano po intervalima: 0, 1, 2
26. Relapse - povratak bolesti: 0 ili 1
27. aGvHDIIIIV - da li se razvila neka bolest 3. i 4. stepena: 0 ili 1
28. <font color=red>extcGvHD - da li se razvila neka bolest 3. i 4. stepena: 0, 1 i ?</font>
29. CD34kgx10d6 numeric - doza neke vrste ćelija skalirana po težini čoveka: float 0.79 - 57.78 
30. <font color=red>CD3dCD34 numeric - doza druge vrste ćelija skalirana po dozi druge vrste: float 0 - 99.56097 i ?</font>
31. <font color=red>CD3dkgx10d8 numeric - doza neke vrste ćelija skalirana po težini čoveka: float 0 - 20.02 i ?</font>
32. <font color=red>Rbodymass numeric - bodymass indeks pacijenta: float 0 - 103.4 i ?</font>
33. <font color=green>ANCrecovery numeric - oporavak neutrofila definisan kao njihov broj (>0.5 je da) : float 9 - 26.0 i 1000000.0 (nisam sigurna da li je ovo neki fuš podatak)</font>
34. <font color=green>PLTrecovery numeric - oporavak trombocita definisam njihov broj (>50000 da): float 9 - 285 i 1000000.0 (isti komentar)</font>
35. <font color=green>time_to_aGvHD_III_IV numeric - vreme u danima proteklo do razvoja akutne bolesti transplantata 3. ili 4. stepena: 10 - 100 i 1000000 (isti komentar)</font>
36. survival_time numeric - vreme posmatranja (u slučaju živog pacijenta) ili vreme do smrti: 6 - 3364
37. survival_status numeric - da li je osoba živa ili mrtva: 0 ili 1


In [829]:
print(set(data.dtypes))
data.dtypes[data.dtypes == 'O']

{dtype('int64'), dtype('O'), dtype('float64')}


RecipientABO           object
RecipientRh            object
ABOmatch               object
CMVstatus              object
DonorCMV               object
RecipientCMV           object
Disease                object
Antigen                object
Alel                   object
extcGvHD               object
CD3dCD34 numeric       object
CD3dkgx10d8 numeric    object
Rbodymass numeric      object
dtype: object

U tabeli se nalaze vrednosti koje nisu samo numeričke ( int / float ).     
Potrebno je proveriti da li neke kolone koje imaju više tipova podataka (njihov tip je object) sadrže i nepostojeće vrednosti (obeležene su sa "?" npr). Kolone koje sadrže neprihvaćene vrednosti su u podacima obeležene crvenom bojom.

#### Sređivanje podataka

Ni u jednoj klasi fičera ne nedostaje veliki broj vrednosti, tako da se sve nevalidne vrednosti menjaju medianom (vrednosti sa nejvećim brojem ponavljanja).

In [830]:
def find_median(col):
    sorted_list = sorted(list(data[col][data[col] != '?']))
    return sorted_list[len(sorted_list)//2]

In [831]:
# zamena vrednosti '?' medianom
data['RecipientABO'] = data['RecipientABO'].replace('?', int(find_median('RecipientABO')))
data['RecipientRh'] = data['RecipientRh'].replace('?', int(find_median('RecipientRh')))
data['ABOmatch'] = data['ABOmatch'].replace('?', int(find_median('ABOmatch')))
data['CMVstatus'] = data['CMVstatus'].replace('?', int(find_median('CMVstatus')))
data['DonorCMV'] = data['DonorCMV'].replace('?', int(find_median('DonorCMV')))
data['RecipientCMV'] = data['RecipientCMV'].replace('?', int(find_median('RecipientCMV')))
data['Antigen'] = data['Antigen'].replace('?', int(find_median('Antigen')))
data['Alel'] = data['Alel'].replace('?', int(find_median('Alel')))
data['extcGvHD'] = data['extcGvHD'].replace('?', int(find_median('extcGvHD')))
data['CD3dCD34 numeric'] = data['CD3dCD34 numeric'].replace('?', float(find_median('CD3dCD34 numeric')))
data['CD3dkgx10d8 numeric'] = data['CD3dkgx10d8 numeric'].replace('?', float(find_median('CD3dkgx10d8 numeric')))
data['Rbodymass numeric'] = data['Rbodymass numeric'].replace('?', float(find_median('Rbodymass numeric')))

Svi podaci koji su brojevi a predstavljeni su kao string u bazi se konvertuju u odgovarajući tip podataka

In [832]:
data['RecipientABO'] = data['RecipientABO'].astype('int64')
data['RecipientRh'] = data['RecipientRh'].astype('int64')
data['ABOmatch'] = data['ABOmatch'].astype('int64')
data['CMVstatus'] = data['CMVstatus'].astype('int64')
data['DonorCMV'] = data['DonorCMV'].astype('int64')
data['RecipientCMV'] = data['RecipientCMV'].astype('int64')
data['Antigen'] = data['Antigen'].astype('int64')
data['Alel'] = data['Alel'].astype('int64')
data['extcGvHD'] = data['extcGvHD'].astype('int64')
data['CD3dCD34 numeric'] = data['CD3dCD34 numeric'].astype('float64')
data['CD3dkgx10d8 numeric'] = data['CD3dkgx10d8 numeric'].astype('float64')
data['Rbodymass numeric'] = data['Rbodymass numeric'].astype('float64')

Kodovanje kategoričkog atributa disease 

In [833]:
print(set(data['Disease']))
data = data.replace({'Disease' : {'chronic' : 0, 'lymphoma' : 1, 'AML' : 2, 'nonmalignant' : 3, 'ALL': 4}})
data['Disease']

{'chronic', 'lymphoma', 'AML', 'nonmalignant', 'ALL'}


0      4
1      4
2      4
3      2
4      0
      ..
182    1
183    2
184    3
185    0
186    3
Name: Disease, Length: 187, dtype: int64

In [834]:
data.dtypes

Recipientgender                   int64
Stemcellsource                    int64
Donorage numeric                float64
Donorage35                        int64
IIIV                              int64
Gendermatch                       int64
DonorABO                          int64
RecipientABO                      int64
RecipientRh                       int64
ABOmatch                          int64
CMVstatus                         int64
DonorCMV                          int64
RecipientCMV                      int64
Disease                           int64
Riskgroup                         int64
Txpostrelapse                     int64
Diseasegroup                      int64
HLAmatch                          int64
HLAmismatch                       int64
Antigen                           int64
Alel                              int64
HLAgrI                            int64
Recipientage numeric            float64
Recipientage10                    int64
Recipientageint                   int64


#### Podatke koji u sklopu nekog karakterističnog obeležja čine manjinu (manje od 5%) je potrebno smestiti u zasebnu kategoriju pod 'ostalo' obeleženo kao numerička vrednost karakteristična za svaku klasu

Za većinu karakteristična obeležja uglavnom se samo jedna vrednost pojavljivala manje od 5%, tako da nije bilo potrebe spajati više klasa.

In [867]:
# provera procentualnoti ponavljanja vrednosti atributa
attr = data.columns[36]
print(attr)
print(set(data[attr]))
val = 0
print(round(len(data[attr][data[attr] == val])/len(data[attr])*100),"%")

survival_status numeric
{0, 1}
55 %


Za karakteristično obeležje "Alel", vrednost "2" se javlja 3%, dok se vrednost "3" javlja 1% - njih je potrebno spojiti u jednu klasu i ostale je potrebno preimenovati

In [814]:
data[data.columns[20]].replace(3, 2, inplace = True)
print(set(data[data.columns[20]]))

{0, 1, 2, -1}


Za karakteristično obeležje "HLAgrI", vrednost "3" se javlja 5%, vrednost "5" javlja 2%, a vrednost "7" se javlja 3% - njih je potrebno spojiti u jednu klasu i ostale je potrebno preimenovati

In [845]:
data[data.columns[21]].replace(5, 3, inplace = True)
data[data.columns[21]].replace(7, 3, inplace = True)
print(set(data[data.columns[21]]))

{0, 1, 2, 3, 4}


# 2. Analiza i izbor obeležja